nion/swift/Panel.py

```diff
@@ -248,13 +251,26 @@ def __init__(self, ui_settings: UISettings.UISettings, title: typing.Optional[st
         self.on_close_clicked: typing.Optional[typing.Callable[[], None]] = None
         self.on_context_menu_clicked: typing.Optional[typing.Callable[[int, int, int, int], bool]] = None
         self.on_double_clicked: typing.Optional[typing.Callable[[int, int, UserInterface.KeyboardModifiers], bool]] = None
+        self.on_edit_clicked: typing.Optional[typing.Callable[[], None]] = None
         self.on_tool_tip: typing.Optional[typing.Callable[[], typing.Optional[str]]] = None
         self.__mouse_pressed_position: typing.Optional[Geometry.IntPoint] = None
+        self.__edit_control_rect: typing.Optional[Geometry.IntRect] = None
+        self.__close_control_rect: typing.Optional[Geometry.IntRect] = None
+        self.__get_font_metrics: typing.Optional[typing.Callable[[str, str], UISettings.FontMetrics]] = None
+
+        if get_font_metrics_fn:
+            self.__get_font_metrics = get_font_metrics_fn
+        elif self.__ui_settings:
+            if hasattr(self.__ui_settings, "ui"):
+                local_ui = self.__ui_settings.ui
+                if hasattr(local_ui, "get_font_metrics"):
+                    self.__get_font_metrics = local_ui.get_font_metrics
```

Comment:
I didn't see this on first pass of review - but this is again breaking encapsulation. In this case, let's just change the HeaderCanvasItem to _not_ take the `get_font_metrics_fn` argument. It's redundant since `UISettings` is required and supplies the same info. Also, we can eliminate the `label` argument too. Unused. Also, the `HeaderCanvasItem` is only used in two places and only one of those two places `FilterPanel.py` requires eliminating the `get_font_metrics_fn` argumnet. Neither of those places passes the `label` argument, so that can just be removed I think.

Reply:
You can use `self.__ui_settings.get_font_metrics(...)`. No need to pass in `get_font_metrics_fn` anymore.